### DISCUSSION TITLE
having a trainable parameter in the reference solution

### DISCUSSION THREAD

**Discussion by Burhan3228** [lululxvi/deepxde#1746] [Q&A]
Dear all, in the code below, i have the reference solution as:

`    def fun_initnewTF(x):
        
        condition1 = tf.greater_equal(x[:, 0:1], 0.0)
        condition2 = tf.greater_equal(xstensor - x[:, 0:1], 0.0)
        condition3 = tf.greater_equal(xstensor + x[:, 0:1], 0.0)
        
        result1 = tf.where(condition2, tf.ones_like(x[:, 0:1]), tf.zeros_like(x[:, 0:1]))
        result2 = tf.where(condition3, tf.ones_like(x[:, 0:1]), tf.zeros_like(x[:, 0:1]))
        
        result =  tf.where(condition1, result1, result2)
        
        return result

    def mycritiinewnewTF(x, theta):
        numerator = 3 * theta * tf.sqrt(2.0)
        denominator = (1 + theta) * tf.sqrt(2.0) + tf.math.cosh(
            x * tf.sqrt(theta)
        ) * tf.sqrt(2 - 5 * theta + 2 * theta ** 2)
        return numerator / denominator

    def mycritiisonnewnewTF(x):
        spatial_component = x[:, 0:1]  # Ensure operation on spatial component only
        return mycritiinewnewTF(spatial_component, theta=0.45)

    def reference_solutionTF(x):
        # x should be a TensorFlow tensor of shape (N, 2) where each row is [space, time]
        spatial_component = x[:, 0:1]  # Extract spatial component
        time_component = x[:, 1:2]     # Extract time component
        # Initialize an output tensor
        result = tf.zeros_like(spatial_component)
        
        # Apply conditions based on the time component
        zero_mask = tf.equal(time_component, 0.0)
        nonzero_mask = tf.not_equal(time_component, 0.0)
        
        result = tf.where(zero_mask, US_trainable*fun_initnewTF(spatial_component), result)
        result = tf.where(nonzero_mask, mycritiisonnewnewTF(x), result)
        
        return result`

Here, US_trainable is my trainable parameter , in other word the reference solution itself depends on the trainable parameter.  When i use it the data as:

`    data = dde.data.TimePDE(
        geomtime,
        pde,
        [bc,ic,observe_y_t_vals_except_inits],
        num_domain=1024,
        num_boundary=16,
        num_initial=256,
        solution=reference_solutionTF,
        anchors=combine_except_inits,
        train_distribution="Sobol",
        num_test=10000
    )`
    
    
    it produces the following error:
    
    TypeError: The value of a feed cannot be a tf.Tensor object. Acceptable feed values include Python scalars, strings, lists, numpy ndarrays, or TensorHandles. For reference, the tensor object was Tensor("Select_24:0", shape=(21768, 1), dtype=float32) which was passed to the argument `feed_dict` with key Tensor("Placeholder_15:0", shape=(None, 1), dtype=float32).

Apparently, it doesn't accept the reference solution in tensorflow environment and it asks its numpy version. Here is the problem: I d like to have US_trainable to be trained during training and the reference solution therefore also need to be adjusted accordingly during training. how can i achieve it? thanks in advance.

**Comment by lululxvi** [comment]
Reference solution cannot have unknown variables.
